SignalServiceKit/src/Messages/OWSMessageSender.m

```diff
@@ -954,11 +954,29 @@ - (void)sendMessage:(TSOutgoingMessage *)message
 
     NSString *localNumber = [TSAccountManager localNumber];
     if ([localNumber isEqualToString:recipient.uniqueId]) {
+        OWSAssert([message isKindOfClass:[OWSOutgoingSyncMessage class]]);
+        // Messages send to the "local number" should be sync messages.
+        //
+        // We can skip sending sync messages if we know that we have no linked
+        // devices. However, we need to be sure to handle the case where the
+        // linked device list has just changed.
+        //
+        // The linked device list is reflected in two separate pieces of state:
+        //
+        // * OWSDevice's state is updated when you link or unlink a device.
+        // * SignalRecipient's state is updated by 409 "Mismatched devices"
+        //   responses from the service.
+        //
+        // If _both_ of these pieces of state agree that there are no linked
+        // devices, then can safely skip sending sync message.
```

Comment:
I think this depends too much about the state of the OWSDevice collection. That collection (in it's current implementation) is not guaranteed to be in sync with the service. e.g. if the user leaves the screen or their app crashes before completing the linking process on desktop, we may not have knowledge of the secondary device. That's not new with this PR, but what's in master as of https://github.com/WhisperSystems/Signal-iOS/commit/1014f27f45fa12645b3cb7c02ea30f36a4ca50f3 is incorrect.

Reply:
I think the approach you mentioned earlier, about setting a dirty flag "mayHaveLinkedDevice" or something when the user initiates the device linking process.